Atualização da API, Implementação de Service worker e IndexedDB, Filtro de busca e contagem de créditos

Samlibardi opened the pull request:
Nota: Abri este PR antes da conclusão das mudanças pretendidas de forma que seja possível o teste no ambiente de homologação e feedback interativo durante o desenvolvimento. As mudanças ainda não realizadas estão destacadas no corpo do PR.

# Crawler + scraper

- Código convertido para Python 3.x
- Dependências agora são gerenciadas usando Pip. Isto torna o código fonte mais limpo pois contém apenas o necessário. Para instalar as dependências, basta ter o pipenv instalado e executar `pipenv install` no diretório do projeto.
- Disciplinas são buscadas paralelamente para melhorar o tempo de execução. Isso permite um debugging mais rápido e eficiente
- Log de execução salvo no diretório através do uso do módulo logger.
- Adicionados parâmetros de entrada (usar o argumento --help ou -h mostra as possibilidades)
- Busca por mais informações das disciplinas, como créditos, unidade, departamento e programa resumido.
  - Informações são salvas em json associativo. Exemplo: 
`{'nome' : 'Cálculo', 'codigo' : 'MAT0141', etc..}`
isso permite maior retrocompatibilidade para versões futuras da API, pois novos campos não afetam a interpretação do cliente.
  - Salva o resultado de duas formas: Arquivos na forma XXXXXXX.json para cada disciplina e db.json (padrão, pode ser mudado pelo argumento -o/--out) que contém todas as disciplinas.
  - Também salva uma versão compactada .gz de cada arquivo no mesmo diretório. Com o auxílio do mod_rewrite do Apache, é possível servir essa versão para navegadores que aceitem arquivos compactados. Isso reduz em **80%** o tamanho do arquivo final, devido à alta repetitividade do json associativo.

# Cliente

 - Adaptações para o uso da nova API
 - Mudança na maneira como é salvo o state. Agora, contém apenas informações básicas: código das disciplinas e turmas selecionadas (#57), última cor utilizada, plano selecionado etc. O metadata da disciplina é obtido do banco de dados ou do servidor, de forma que as informações são atualizadas entre sessões.
 - Acrescentado contador de créditos da grade atual (#44)
 - Exportar para calendário: Acrescentado campo "observação" da turma como descrição do evento. Em alguns casos, esse campo possui informações como local de aula, o que pode ser útil. 
 - Adicionado banner abaixo do cabeçalho para mensagens do sistema, para diminuir a intrusividades de mensagens enviadas usando `alert()` (#38)
 - Cache agora é gerenciado por um Service Worker. A estratégia de cache utilizada foi baseada na [cache-update-refresh](https://serviceworke.rs/strategy-cache-update-and-refresh_service-worker_doc.html) de maneira a otimizar o tempo de carregamento da página. A atualização é feita on-demand com o uso de ETags. Em caso de atualização obtida, um banner requisitando o refresh da página é mostrado. O banco de dados não é coberto por esse cache.
 - Atualização do banco de dados local e busca é feita por web workers. Como web workers funcionam em threads separados, a performance de carregamento obteve uma melhora significativa (havia um bloqueio de UI durante o carregamento do banco de dados)
   - Banco de dados agora é salvo e utilizado usando IndexedDB ao invés de ficar salvo na memória. A IndexedDB é perpetuada entre sessões e renovada no início da sessão. A ETag da última atualização é salva e enviada durante o pedido HTTP, e o banco de dados só é baixado caso haja alguma alteração.
   -  O carregamento dos dados para o IndexedDB tem feedback visual por meio de uma barra de carregamento é mostrada no cabeçalho da página. Isso dá mais clareza ao usuário sobre o funcionamento da página.
 - Acrescentado AppManifest e ícone para uso offline em sistemas Android. [WIP: melhorar o ícone]
 - Implementada busca avançada com filtragem por **campus**, unidade, departamento e horário (#3)

Samlibardi committed:
Added web workers to update and search database
Lecture and trigram info now stored in IndexedDB instead of memory. Update worker will fetch json and update accordingly. Search worker will read IndexedDB in separate thread and message search result.

Samlibardi committed:
Fixed "undefined" problems in trigrams db

Samlibardi committed:
Fixed smallcase code search and db update if indexeddb already exists

Samlibardi committed:
Removed whole string "trigram"
Whole string attribute doesn't contribute much to search result but takes a large amount of space in DB. It's not included anymore

Samlibardi committed:
Only one dbsearch worker runs for searchbox now.
This worker caches last searches and trigrams in memory for faster retrievals. Queries run in <5ms usually

Samlibardi committed:
DocumentFragment to add lectures to searchbox
Lectures are now appended to a document fragment that is then appended to the searchbox, reducing reflows to only one and improving performance specially in Edge

Samlibardi committed:
Set IndexedDB name to constant defined in dbhelpers.js

Samlibardi committed:
Service Worker added, AppCache removed
Caching is now managed by service worker instead of AppCache. AppCache was removed.

Samlibardi committed:
Added header loading-bar for db loading status

Samlibardi committed:
Worker fixes

Samlibardi committed:
Firefox fix

Samlibardi committed:
Redirect to old version if Web Workers is not supported

Samlibardi committed:
Removed old database code

Samlibardi committed:
Database is now cleared before downloaded data is saved
Prevents deleted lectures to show as current

Samlibardi committed:
State loading updated to Fetch API from XMLHttpRequest

Samlibardi committed:
State now loads lecture info from DB on load
This means outdated states will now have updated info as soon as the page loads

Samlibardi committed:
Added banner to display user-friendly messages.
Banner will replace alert for identifier saving/loading responses, page update available and error messages.

Samlibardi committed:
Added WebApp manifest for offline mobile usage

Samlibardi committed:
Add icons and WebManifest

Samlibardi committed:
Properly end db loader when db fails to download from server

Samlibardi committed:
New crawler
Rebuilt in Python 3, async HTTP fetch, associative json result (new db schema) with more lecture info

Samlibardi committed:
Fixed nonstandard lecture code parsing

Samlibardi committed:
Added optional arguments for output file and compression

Samlibardi committed:
Client support to new schema. Calenadar now shows extra classrom info. Added credit counter.

Samlibardi committed:
Ignore lectures without valid scheduled times.

Samlibardi committed:
Added pipfiles to install dependencies

Samlibardi committed:
Exception fixes

Samlibardi committed:
Added Dexie for advanced search queries

Samlibardi committed:
Filtered search implemented.
User is now able to select unit and department to get lectures from.

Samlibardi committed:
Fixed sunday classroom causing exception

Samlibardi committed:
Fixed classroom deletion exception

Samlibardi committed:
Added styling to new elements

Samlibardi committed:
Fixed unit field population on first-time startup and no-result search box styling

Samlibardi committed:
Fixed file uploading and downloading

Samlibardi committed:
Search options now perpetuate between sessions

Samlibardi committed:
Improved search option boxes.

Samlibardi committed:
Better credit display styling

Samlibardi committed:
Fixed showing less than all lectures on state load

Samlibardi committed:
Added campus filtering

Samlibardi committed:
Fixed credit display when no lectures on plan

pedropaulovc commented:
Olá Samuel, esse PR está ficando bem grande... Me avise quando você achar que ele está pronto para o code review.

Samlibardi commented:
Certo. Acho que algumas mudanças são modulares o suficiente para serem implementadas em um PR separado.
Por enquanto, acho que já tenho um MVP disponível para teste, se quiser já começar os testes no ambiente de homologação. 

Vou começar a limpeza e documentação do código e polimento da interface e acredito que até semana que vem o PR estará pronto para review.

Tenho algumas ressalvas, no entanto. Como o Edge/IE não tem suporte a índices compostos no Indexeddb, a performance da busca filtrada ficou bastante comprometida, e embora a busca normal tenha performance aceitável, esse browser não tem realmente nenhuma vantagem em usar a implementação do DB em indexeddb. Minha dúvida é se vale a pena fazer um fallback gracioso e esconder o filtro de busca nessa condição ou se é melhor redirecionar o usuário para a versão atual, semelhante ao que acontece para o matrusp v1

Samlibardi committed:
Fixed #36

Samlibardi committed:
Fixed added lectures not selected

Samlibardi committed:
Fixed all plans showing as active when loading state

Samlibardi committed:
Improved search options summary text

Samlibardi committed:
Added lecture unit and department to search-box results

Samlibardi committed:
Improved filtering logic

Samlibardi committed:
Added timeframe filtering

Samlibardi committed:
Client-side linked theoretics and practices logic

Samlibardi committed:
Fixed PDF not displaying classroom codes

Samlibardi committed:
Fixed selected unit being deselected on campus changed

Samlibardi committed:
Fixed DB not updating when page is refreshed before loading is finished

Samlibardi committed:
Improved search options summary text

Samlibardi committed:
Fixed combination track not updating on change

Samlibardi committed:
Added search options form styling
Also improved summary text generation further

Samlibardi committed:
Fixed credit display not resetting when opening empty plan

Samlibardi committed:
Improved search-box behaviour
Typing now closes options form, clear button also clears search-box text

Samlibardi committed:
Summary text generator improvements

Samlibardi committed:
Added redirection rules.
Redirect to old version if used features are not available in user browser

Samlibardi committed:
Added gracious fallback for browser's lack of complex keys support
Disable certain options that are only possible with the use of compound or multi-key indexes and display a message

Samlibardi committed:
Fixed clear button persisting after adding a lecture

Samlibardi commented:
Agora o PR já está feature-complete. Faltam apenas a limpeza do código e documentação, mas acredito que os testes de homologação já podem ser iniciados sem problemas antes do code-review. 

Para executar no servidor:
Usando o python3, 
 - `pip install pipenv` se ainda não estiver instalado,
 - `pipevn install` para instalar as dependências do projeto
 - `pipenv run py parse_usp.py ..\db` ou `pipenv shell` + `py parse_usp.py ..\db` para executar o código com as dependências. O parâmetro ..\db gerará os arquivos no diretório configurado no cliente. Os parâmetros opcionais podem ser consultados usando `-h`

Pela nomenclatura do arquivo do script e dos arquivos gerados, há compatibilidade com o código anterior, mas como o anterior apaga todos os arquivos da pasta e abre todos os arquivos para parsear, é necessário que o **script novo seja executado após o término do anterior**,  **ou** utilizar **outro diretório no anterior**.

Após a execução do script no servidor, é possível utilizar o cliente. Ele possui compatibilidade com states da versão atual, portanto é possível fazer upload e carregar identificadores salvos na versão atual. 
Testes sugeridos:
 - Buscar e adicionar disciplinas usando os filtros
 - Observar a leitura de créditos da grade conforme alterações
 - Testar o cliente offline (funciona apenas no Chrome, Firefox, Edge 17+ e Safari 11.1)
 - Testar deployments n+1 (fazer pequenas modificações no código e verificar se as mudanças são recebidas)

Samlibardi committed:
Search updates on options changed + css-based result box visibility

Samlibardi committed:
Improved NLG and visual feedbacks

pedropaulovc commented:
Samuel, eu carreguei o código do commit 7551076b10810c504098550 no ambiente de homologação. Eu tentei montar o ambiente para rodar o scraper python mas encontrei um problema: o servidor do MatrUSP roda Debian Jessie, então seria ótimo que o script fosse compatível com Python 3.4. Eu tentei ajustar o Pipfile manualmente mas não tive muito sucesso. Também parece que o código do `parse_usp.py` teria que ser modificado nesse caso.

Você acha que seria muito complicado ajustar esse PR para funcionar com Python 3.4?

Samlibardi committed:
Fixed lecture not selectable in result box

Samlibardi commented:
O código atualmente utiliza a sintaxe async/await que foi introduzida no Python 3.5. Nesse caso, é possivel trocar para utilizar o decorador @asyncio.coroutine/yield from que é compatível com Python 3.4.
Como eu fiz o código em 3.6, não tenho certeza se utilizei mais algum recurso que não tem suporte nas versões mais antigas.

________________________________
From: Pedro Paulo <notifications@github.com>
Sent: Monday, April 16, 2018 12:36:03 AM
To: matrusp/matrusp
Cc: Samuel Libardi Godoy; Author
Subject: Re: [matrusp/matrusp] [WIP] Atualização da API, Implementação de Service worker e IndexedDB, Filtro de busca e contagem de créditos (#75)


Samuel, eu carreguei o código do commit 7551076<https://github.com/matrusp/matrusp/commit/7551076b10810c50409855087714ac90876a95db> no ambiente de homologação. Eu tentei montar o ambiente para rodar o scraper python mas encontrei um problema: o servidor do MatrUSP roda Debian Jessie, então seria ótimo que o script fosse compatível com Python 3.4. Eu tentei ajustar o Pipfile manualmente mas não tive muito sucesso. Também parece que o código do parse_usp.py teria que ser modificado nesse caso.

Você acha que seria muito complicado ajustar esse PR para funcionar com Python 3.4?

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/matrusp/matrusp/pull/75#issuecomment-381471096>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AcMGs2pltedJoNtTsJsMTFtyPwbWqjTtks5tpBGjgaJpZM4S0dvB>.

Samlibardi committed:
Fixed campus selection not persisting

Samlibardi commented:
Como esse script depende do [aiohttp](https://pypi.python.org/pypi/aiohttp) para funcionar, e este requer Python 3.5.3 ou maior, infelizmente não teria como converter o código para compatibilizar sem reescrever boa parte do funcionamento dele.  O máximo que é possível fazer neste caso é retornar do 3.6 ao 3.5 transformando as f-strings.
Se não for possível atualizar o Python da máquina, posso procurar alguma solução, talvez utilizando uma versão antiga do aiohttp que funcione no 3.4, mas não sei se existe.

pedropaulovc commented:
Vou estudar se há alguma forma de instalar o Python 3.6 no Debian apenas para o usuário que vai rodar o script. O que eu não queria ter que fazer é habilitar o Debian Testing no sources.config.

Samlibardi commented:
https://unix.stackexchange.com/a/332658

É possível compilar o python diretamente do código fonte, segundo essa resposta. Segue o código sugerido (atualizado para a rev mais recente):
```
wget https://www.python.org/ftp/python/3.6.5/Python-3.6.5.tgz
tar xvf Python-3.6.5.tgz
cd Python-3.6.5
./configure --enable-optimizations
make -j8
sudo make altinstall
python3.6
```

Não tenho muito conhecimento em Linux, então não sei dizer quais implicações esse método teria. Espero que funcione

Samlibardi committed:
Fixed AppCache not properly disabled

Samlibardi committed:
Fixed button text selectable

Samlibardi committed:
Code cleanup and documentation

Samlibardi commented:
Atualizei a documentação e fiz a limpeza do código. Acredito que agora o PR esteja pronto para revisão.

Samlibardi committed:
Fixed missing cached file

Samlibardi committed:
Enable identifier caching

pedropaulovc commented:
> Acredito que agora o PR esteja pronto para revisão.

Maravilha! Só peço paciência até que eu revise tudo. Esse final de semana eu estarei em viagem e não terei tempo para trabalhar nesse PR. Espero fazer algum progresso na quarta-feira à noite e com sorte terminar no próximo final de semana.

pedropaulovc commented:
Como estamos no período entre matrículas eu decidi mandar estas mudanças para produção. Se eu encontrar algum problema, eu abrirei bugs separadamente. Minhas desculpas pela demora.

pedropaulovc commented:
Ah, e esse pull request merece um post no Facebook! A única dúvida é se a gente deveria esperar até chegarmos no período de matrículas ou postar agora.

coelhodepina commented:
O período de matrículas é de 25 de junho a 10 de julho.
Por volta do meio de junho seria realmente legal se vocês divulgassem no Facebook todas as melhorias que fizeram.

Samlibardi commented:
Que bom que o deploy foi executado com sucesso. Acho que, como ainda faltam quase 2 meses para as matrículas, dá tempo de fazer algumas pequenas melhorias antes de fazer a divulgação. Eu diria que do início pro meio de junho seria um período legal.
